hashintel/hash
# Implement BP hooks for creating/updating/deleting linked aggregations

## 🌟 What is the purpose of this PR?




This PR implements three new BP methods for creating/updating/deleting the linked aggregations of an entity.

## ⚠️ Known issues



- There is a bug presumed to be related to the entity store where updates to linked aggregations are no longer reflected in the entity store. This means updating an aggregation (when toggling the entity type displayed in the table block for example) does not get reflected in the block, although the update was saved in the datastore.

## 🚫 Blocked by





## 🐾 Next steps



- [update the `blockprotocol` package to include the updated BP method types](https://app.asana.com/0/1201095311341924/1202216335609392/f)
- [update the BP spec to include the introduced BP aggregation methods](https://app.asana.com/0/1201095311341924/1202216335609402/f)
- [implement the `useBlockProtocolUpdateLinks` hook to update links using their `linkId`](https://app.asana.com/0/1201095311341924/1202216335609406/f)

## 🔍 What does this change?





#### GQL API

- updates the `deleteLinkedAggregation` and `updateLinkedAggregationOperation` GQL mutations to use `aggregationId` as an identifier

#### Blockprotocol

- patches the `blockprotocol` NPM package to include `BlockProtocolCreateLinkedAggregationsFunction`, `BlockProtocolUpdateLinkedAggregationsFunction` and `BlockProtocolDeleteLinkedAggregationsFunction` type definitions

#### Frontend

- implements `useBlockProtocolCreateLinkedAggregations`, `useBlockProtocolUpdateLinkedAggregations` and `useBlockProtocolDeleteLinkedAggregations` hooks
- updates the Table block to make use of the new BP methods

### 📜 Does this require a change to the docs?



No (updating the BP spec will be handled in a separate task)

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1201095311341924/1201866327702956/f) _(internal)_

## 🛡 What tests cover this?



- integration tests cover create/update/delete linked aggregation GQL mutations

## ❓ How to test this?



1. Build & serve the table block locally
2. Login to the workspace, load the locally hosted table block
3. Modify the entity type displayed in the table to update the aggregation. Note that nothing changes because of am existing bug in the entity store.
4. Edit another block in the page, to refresh the entity store. The block will now be updated with the new linked aggregation, and display the entity type you previously selected.

## 📹 Demo

https://user-images.githubusercontent.com/42802102/164500974-c95fff65-bbc2-4545-974a-70122027df7e.mov
